### PR TITLE
HDDS-6722. Ensure RocksObjects are closed.

### PR DESCRIPTION
**Pull request by duongnguyen0**
## What changes were proposed in this pull request?

I scanned all the usages of RocksObject and ensure they are closed.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6724
https://issues.apache.org/jira/browse/HDDS-6726

## How was this patch tested?

Standard CI: https://github.com/duongnguyen0/ozone/actions/runs/2829235416

I ran some workload (with new container allocation) on localhost and verified no object leak in OM, SCM and DNs.  
Screen Shot 2022-08-10 at 8 20 13 AM

### CODE REVIEW

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBSstFileWriter.java [hunk 1]
```diff
@@ -82,19 +82,23 @@ public void close() throws IOException {
...
-        emptyOption.close();
+        closeResources();
       }
 
       keyCounter.set(0);
     }
   }
 
+  private void closeResources() {
+    sstFileWriter.close();
```
Comment (kerneltime) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBSstFileWriter.java:93]: Check if not null.

**Comment by kerneltime** [comment]
@duongnguyen0 thank you for the quick turn around on this. The report for which RocksObject were used was generated using IntelliJ?

**Comment by duongnguyen0** [comment]
> @duongnguyen0 thank you for the quick turn around on this. The report for which RocksObject were used was generated using IntelliJ?

It's done manually. I went over the list of RocksObject subclasses to see how each one is used.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBSstFileWriter.java (hunk 1 above)
Comment (duongnguyen0) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBSstFileWriter.java:93]: Null check is done before every usage of this method.

**Comment by kerneltime** [comment]
@ChenSammi @jojochuang @smengcl Can you please take a look?
@ChenSammi do add anyone you know who can review this PR, please feel free to add them.
Post this merge, we should be able to upgrade RocksDB Version. It is a critical PR that should help stem any memory leaks that are hard to debug with RocksDB.

**Comment by szetszwo** [comment]
@duongnguyen0 , thanks for working on this!  The changes here are non-trivial and very big (73KB).   Could you split this into multiple JIRAs/pull requests?  The first one could be `DBProfile` related.  I am looking at it but it is not easy to understand.

BTW, each pull request should refer to a unique JIRA so that it is easy to track.  Ideally, the JIRA-to-pull-request mapping should be 1-to-1.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedColumnFamilyOptions.java [hunk 2]
```diff
@@ -36,6 +36,10 @@ public ManagedColumnFamilyOptions(ColumnFamilyOptions columnFamilyOptions) {
   @Override
   public ManagedColumnFamilyOptions setTableFormatConfig(
       TableFormatConfig tableFormatConfig) {
+    // Close the previous TableFormatConfig before overwriting.
+    if (tableFormatConfig() instanceof ManagedBlockBasedTableConfig) {
+      ((ManagedBlockBasedTableConfig) tableFormatConfig()).close();
+    }
```
Comment (szetszwo) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedColumnFamilyOptions.java:42]: It is better to throw an exception for overwriting.  We may throw an UnsupportedOperationException.
```
    final TableFormatConfig previous = tableFormatConfig();
    if (previous != null) {
      throw new UnsupportedOperationException("Overwrite is not supported");
    }
```
Or we may enforce close before overwrite.
```
    if (previous instanceof ManagedBlockBasedTableConfig) {
      if (!previous.isClose()) { // isClose() is a new method
        throw new IllegalArgumentException("Overwriting an unclosed value");
      }
    } else if (previous != null) {
      throw new UnsupportedOperationException("Overwrite is not supported for " + previous.getClass());
    }
```

**Review by szetszwo — COMMENTED** [review]
@duongnguyen0 , we probably should not try to close the overwritten objects -- they should be closed before overwrite.  Or we may throw an exception and choose to not supporting overwrite since it does not seem useful.  See also the comment inlined.

**Comment by duongnguyen0** [comment]
> @duongnguyen0 , thanks for working on this! The changes here are non-trivial and very big (73KB). Could you split this into multiple JIRAs/pull requests? The first one could be `DBProfile` related. I am looking at it but it is not easy to understand.
> 
> BTW, each pull request should refer to a unique JIRA so that it is easy to track. Ideally, the JIRA-to-pull-request mapping should be 1-to-1.

Yeah, I feel the same after review it myself. I'm marking this as Draft PR and splitting it to smaller ones.

**Comment by duongnguyen0** [comment]
Remaining of this PR has been put in #3678. Closing this one.

**Pull request closed without merging**